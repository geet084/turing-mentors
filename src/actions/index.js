export const isLoading = (bool) => ({
  type: 'IS_LOADING',
  isLoading: bool
});

export const hasErrored = (message) => ({
  type: 'HAS_ERRORED',
  message
});

export const updateForm = (form) => ({
  type: 'UPDATE_FORM',
  form
});

export const resetForm = () => ({
  type: 'RESET_FORM',
})

export const getMentorsSuccess = (mentors) => ({
  type: 'GET_MENTORS_SUCCESS',
  mentors
});

export const slackSuccess = (message) => ({
  type: 'SLACK_SUCCESS',
  message
});

export const postMentorSuccess = (mentor) => ({
  type: 'POST_MENTOR_SUCCESS',
  mentor
});