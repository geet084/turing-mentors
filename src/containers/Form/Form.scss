.complete {
  h1 {
     margin: 25px 10px; 
  font-size: 3rem;
  color: #ccc;
  }
  a {
    margin: 20px 8px;
    font-size: .95rem;
    text-decoration: none;
    color: #5f5f5f;
    font-weight: 600;
    &:hover {
      margin-bottom: -3px;
      border-bottom: 3px solid #12f1fc;
      cursor: pointer;
    }
  }
}