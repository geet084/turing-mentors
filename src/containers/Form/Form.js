import React, { Component } from 'react';
import UserInfo from '../../components/UserInfo/UserInfo';
import UserBio from '../../components/UserBio/UserBio';
import UserBackground from '../../components/UserBackground/UserBackground';
import UserSchedule from '../../components/UserSchedule/UserSchedule';
import UserTechSkills from '../../components/UserTechSkills/UserTechSkills';
import UserNonTechSkills from '../../components/UserNonTechSkills/UserNonTechSkills';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { resetForm, updateForm } from '../../actions';
import { sendForm } from '../../thunks/sendForm';

export class Form extends Component {
  state = {
    userInfo: {},
    userBio: {},
    userBackground: {},
    userSchedule: {},
    userTechSkills: [],
    userNonTechSkills: [],
    currentSection: 'userInfo',
  }

  updateUserInfo = (info) => {
    const updatedSection = this.state.currentSection;
    const currentSection = info.pop();
    const data = info.pop()

    this.props.updateForm({ ...this.props.form, ...data })
    this.setState({ [updatedSection]: data, currentSection })

    if (currentSection === 'complete') {
      this.submitForm({ ...this.props.form, ...data })
    }
  }

  submitForm = (form) => {
    const corsPrefix = 'https://cors-anywhere.herokuapp.com/';
    const root = 'https://turing-mentors-be.herokuapp.com';
    const path = `/api/v1/mentors`;
    const url = corsPrefix + root + path;
    this.props.sendForm(url, form)
  }

  reset = () => {
    this.setState({
      userInfo: {},
      userBio: {},
      userBackground: {},
      userSchedule: {},
      userTechSkills: [],
      userNonTechSkills: [],
      currentSection: 'userInfo',
    });
  }

  render() {
    const { currentSection, userInfo, userBio, userBackground, userSchedule, userTechSkills, userNonTechSkills } = this.state;
    const mentor = this.props.location.pathname === '/mentor';
    const user = mentor ? 'Mentor' : 'Mentee';

    return (
      <div>
        {currentSection === 'userInfo' &&
          <UserInfo updateUserInfo={this.updateUserInfo} user={user} {...userInfo} />}
        {currentSection === 'userBio' &&
          <UserBio updateUserInfo={this.updateUserInfo} user={user} {...userBio} />}
        {currentSection === 'userBackground' &&
          <UserBackground updateUserInfo={this.updateUserInfo} user={user} {...userBackground} />}
        {currentSection === 'userSchedule' &&
          <UserSchedule updateUserInfo={this.updateUserInfo} user={user} {...userSchedule} />}
        {mentor && currentSection === 'userTechSkills' &&
          <UserTechSkills updateUserInfo={this.updateUserInfo} user={user} {...userTechSkills} />}
        {mentor && currentSection === 'userNonTechSkills' &&
          <UserNonTechSkills updateUserInfo={this.updateUserInfo} user={user} {...userNonTechSkills} />}
        {currentSection === 'complete' &&
          <div className="complete"><h1>Thank you!</h1><Link to='/' onClick={this.reset}>Return Home</Link> </div>}
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  form: state.form,
})

export const mapDispatchToProps = (dispatch) => ({
  updateForm: (form) => dispatch(updateForm(form)),
  sendForm: (url, form) => dispatch(sendForm(url, form)),
  resetForm: () => dispatch(resetForm()),
})

export default connect(mapStateToProps, mapDispatchToProps)(Form);