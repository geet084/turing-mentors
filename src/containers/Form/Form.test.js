import React from 'react';
import { shallow } from 'enzyme';
import { Form, mapStateToProps, mapDispatchToProps } from './Form';
import { updateForm } from '../../actions';

describe('Form', () => {
  let wrapper;
  let mockProps = {
    location: {
      pathname: '/mentor'
    }
  }

  describe('snapshots', () => {
    beforeEach(() => {
      wrapper = shallow(<Form {...mockProps}/>);
    });

    it('should have match a page one snapshot(default)', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should have match a page two snapshot', () => {
      wrapper.setState({ currentSection: 'userBio' });
      expect(wrapper).toMatchSnapshot();
    });

    it('should have match a page two snapshot', () => {
      wrapper.setState({ currentSection: 'userBackground' });
      expect(wrapper).toMatchSnapshot();
    });

    it('should have match a page three snapshot', () => {
      wrapper.setState({ currentSection: 'userSchedule' });
      expect(wrapper).toMatchSnapshot();
    });

    it('should have match a page four snapshot', () => {
      wrapper.setState({ currentSection: 'complete' });
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('mapStateToProps', () => {
    it('should return an object with all form details', () => {
      const mockState = { firstName: '', lastName: '' };

      const expected = {};

      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call updateForm dispatch', () => {
      const mockDispatch = jest.fn()
      const mockForm = { firstName: 'name', lastName: 'last name' }

      const actionToDispatch = updateForm(mockForm)

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.updateForm(mockForm)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
});