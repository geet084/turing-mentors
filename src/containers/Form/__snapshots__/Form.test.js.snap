// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form snapshots should have match a page four snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Form
    location={
      Object {
        "pathname": "/mentor",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        false,
        false,
        false,
        false,
        <div>
          <h1>
            Thank you!
          </h1>
          <Link
            to="/"
          >
            Return Home
          </Link>
           
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      false,
      false,
      false,
      false,
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1>
              Thank you!
            </h1>,
            <Link
              to="/"
            >
              Return Home
            </Link>,
            " ",
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Thank you!",
            },
            "ref": null,
            "rendered": "Thank you!",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Return Home",
              "to": "/",
            },
            "ref": null,
            "rendered": "Return Home",
            "type": [Function],
          },
          " ",
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          false,
          false,
          false,
          false,
          <div>
            <h1>
              Thank you!
            </h1>
            <Link
              to="/"
            >
              Return Home
            </Link>
             
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        false,
        false,
        false,
        false,
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1>
                Thank you!
              </h1>,
              <Link
                to="/"
              >
                Return Home
              </Link>,
              " ",
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Thank you!",
              },
              "ref": null,
              "rendered": "Thank you!",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Return Home",
                "to": "/",
              },
              "ref": null,
              "rendered": "Return Home",
              "type": [Function],
            },
            " ",
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Form snapshots should have match a page one snapshot(default) 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Form
    location={
      Object {
        "pathname": "/mentor",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <UserInfo
          updateUserInfo={[Function]}
          user="Mentor"
        />,
        false,
        false,
        false,
        false,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "updateUserInfo": [Function],
          "user": "Mentor",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
      false,
      false,
      false,
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <UserInfo
            updateUserInfo={[Function]}
            user="Mentor"
          />,
          false,
          false,
          false,
          false,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "updateUserInfo": [Function],
            "user": "Mentor",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        false,
        false,
        false,
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Form snapshots should have match a page three snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Form
    location={
      Object {
        "pathname": "/mentor",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        false,
        false,
        <UserSchedule
          updateUserInfo={[Function]}
          user="Mentor"
        />,
        false,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      false,
      false,
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "updateUserInfo": [Function],
          "user": "Mentor",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          false,
          false,
          <UserSchedule
            updateUserInfo={[Function]}
            user="Mentor"
          />,
          false,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        false,
        false,
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "updateUserInfo": [Function],
            "user": "Mentor",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Form snapshots should have match a page two snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Form
    location={
      Object {
        "pathname": "/mentor",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        <UserBio
          updateUserInfo={[Function]}
          user="Mentor"
        />,
        false,
        false,
        false,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "updateUserInfo": [Function],
          "user": "Mentor",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
      false,
      false,
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          <UserBio
            updateUserInfo={[Function]}
            user="Mentor"
          />,
          false,
          false,
          false,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "updateUserInfo": [Function],
            "user": "Mentor",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        false,
        false,
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Form snapshots should have match a page two snapshot 2`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Form
    location={
      Object {
        "pathname": "/mentor",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        false,
        <UserBackground
          updateUserInfo={[Function]}
          user="Mentor"
        />,
        false,
        false,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      false,
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "updateUserInfo": [Function],
          "user": "Mentor",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
      false,
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          false,
          <UserBackground
            updateUserInfo={[Function]}
            user="Mentor"
          />,
          false,
          false,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        false,
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "updateUserInfo": [Function],
            "user": "Mentor",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        false,
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
